File: server/lib/key_validator.dart
// server/lib/key_validator.dart
// Mirrors the offline key validation logic from lib/utils/license_manager.dart.
// Must stay in sync with LicenseManager._decodeKey / _generateChecksum.

import 'dart:convert';
import 'package:crypto/crypto.dart';

class KeyValidator {
  final String _secretKey;

  KeyValidator(this._secretKey);

  /// Returns decoded key metadata, or null if the key has an invalid checksum.
  Map<String, dynamic>? decode(String licenseKey) {
    try {
      final cleaned = licenseKey.replaceAll('-', '').toUpperCase();
      if (cleaned.length != 20) return null;
      if (!cleaned.startsWith('CNRI')) return null;
      if (!RegExp(r'^[A-Z0-9]+$').hasMatch(cleaned)) return null;

      final rest = cleaned.substring(4); // strip 'CNRI'
      final typeCode = rest[0];
      final checksum = rest.substring(1, 9); // 8-char HMAC-SHA256 prefix
      final encodedData = rest.substring(9); // 7-char random suffix

      if (_checksum(typeCode + encodedData) != checksum) return null;

      String type;
      DateTime? expiration;
      switch (typeCode) {
        case 'S':
          type = 'standard';
          expiration = DateTime.now().toUtc().add(const Duration(days: 365));
          break;
        case 'P':
          type = 'professional';
          break;
        case 'E':
          type = 'enterprise';
          break;
        default:
          return null;
      }

      return {
        'type': type,
        'expiration': expiration?.toIso8601String(),
      };
    } catch (_) {
      return null;
    }
  }

  String _checksum(String data) =>
      Hmac(sha256, utf8.encode(_secretKey))
          .convert(utf8.encode(data))
          .toString()
          .substring(0, 8)
          .toUpperCase();

  /// Maximum concurrent machine activations per license type.
  static int maxActivations(String type) {
    switch (type) {
      case 'standard':
        return 2;
      case 'professional':
        return 1;
      case 'enterprise':
        return 10;
      default:
        return 1;
    }
  }
}

File: server_php/key_validator.php
<?php
// key_validator.php — Mirrors LicenseManager._decodeKey / _generateChecksum in Dart.
// Must stay in sync with lib/utils/license_manager.dart.

require_once __DIR__ . '/config.php';

/**
 * Validate a license key's HMAC checksum.
 * Returns ['type' => ..., 'expiry' => ...] on success, null on failure.
 */
function validateKeyChecksum(string $key): ?array {
    $cleaned = strtoupper(str_replace('-', '', $key));

    if (strlen($cleaned) !== 20)                   return null;
    if (substr($cleaned, 0, 4) !== 'CNRI')         return null;
    if (!preg_match('/^[A-Z0-9]+$/', $cleaned))    return null;

    $rest       = substr($cleaned, 4);   // 16 chars
    $typeCode   = $rest[0];              // 1 char
    $checksum   = substr($rest, 1, 8);   // 8-char HMAC prefix
    $randomData = substr($rest, 9);      // 7-char random suffix

    // Recompute: HMAC-SHA256(LICENSE_SECRET, typeCode + randomData), first 8 hex chars
    $expected = strtoupper(substr(
        hash_hmac('sha256', $typeCode . $randomData, LICENSE_SECRET),
        0, 8
    ));

    if (!hash_equals($expected, $checksum)) return null;

    switch ($typeCode) {
        case 'S':
            return ['type' => 'standard',     'expiry' => date('Y-m-d H:i:s', strtotime('+365 days'))];
        case 'P':
            return ['type' => 'professional', 'expiry' => null];
        case 'E':
            return ['type' => 'enterprise',   'expiry' => null];
        default:
            return null;
    }
}

/**
 * Maximum concurrent machine activations per license type.
 * Must match KeyValidator.maxActivations() in server/lib/key_validator.dart.
 */
function maxActivations(string $type): int {
    return match ($type) {
        'standard'     => 2,
        'professional' => 1,
        'enterprise'   => 10,
        default        => 1,
    };
}
